### ISSUE TITLE
No eviction with asMap.keySet.toArray

### ISSUE THREAD

**Issue by zzrough** [ben-manes/caffeine#472]
hey,

thanks for the great library.

I'm using it for a long time, but recently I came up with something strange regarding expireAfterWrite eviction.
I don't get exactly get the semantics of the Map view (asMap), since it can or can not be up to date like the API docs tells.

But still, I've got this issue in production that puzzles me:

```
    public static void main(String[] args) {
        Object PRESENT = new Object();

        Cache<String,Object> cache = Caffeine.newBuilder()
                                             .expireAfterWrite(Duration.ofSeconds(3))
                                             .build();

        Consumer<Long> sleepQuietly = millis -> {
            try {
                Thread.sleep(millis);
            } catch (InterruptedException e) {
            }
        };

        new Thread(() -> {
            String name = "u1";

            for (int i = 0; i < 10; i++) {
                cache.put(name, PRESENT);
                // System.out.println(name + ": crowd = " + List.copyOf(cache.asMap().keySet()));            // ERR
                System.out.println(name + ": crowd = " + cache.asMap().keySet().stream().collect(toList())); // OK
                sleepQuietly.accept(500L);
            }
        }).start();

        new Thread(() -> {
            String name = "u2";

            sleepQuietly.accept(1000L);
            cache.put(name, PRESENT);
            // System.out.println(name + ": crowd = " + List.copyOf(cache.asMap().keySet()));            // ERR
            System.out.println(name + ": crowd = " + cache.asMap().keySet().stream().collect(toList())); // OK
        }).start();
    }
````

Simply put:
* eviction is after write 3s
* u1 thread put every 500ms
* u2 thread put after 1s

If I use List.copyOf that relies on toArray() on the key set view, u1 never sees u2 being evicted:
```
u1: crowd = [u1]
u1: crowd = [u1]
u2: crowd = [u1, u2]
u1: crowd = [u1, u2]
u1: crowd = [u1, u2]
u1: crowd = [u1, u2]
u1: crowd = [u1, u2]
u1: crowd = [u1, u2]
u1: crowd = [u1, u2]
u1: crowd = [u1, u2]
u1: crowd = [u1, u2]
```
If the set is traversed using a stream, or the iterator in anyway, I've got the proper eviction:

```
u1: crowd = [u1]
u1: crowd = [u1]
u2: crowd = [u1, u2]
u1: crowd = [u1, u2]
u1: crowd = [u1, u2]
u1: crowd = [u1, u2]
u1: crowd = [u1, u2]
u1: crowd = [u1, u2]
u1: crowd = [u1, u2]
u1: crowd = [u1]
u1: crowd = [u1]
```

Is this normal that keySet iteration works but not the toArray method? If I use List.copyOf or new ArrayList this does not work.
I certainly was not expected this :)

thanks a lot,

**Comment by zzrough** [comment]
(calling cleanUp manually each time did not do a difference)

**Comment by ben-manes** [comment]
Thanks for the test case. It looks like this was optimized to use the backing CHM's faster `toArray`, but missed that it created an inconsistency by not checking for expiration and only iterating if `weakKeys` is enabled. The `values` and `entrySet` iterate, but since `keySet` might be optimizable it has this overload which contains that oversight. Guava always iterates, so we should be consistent.

https://github.com/ben-manes/caffeine/blob/8c91f6544f8f7cbfa5275aa05cce0713ea7f2863/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L2792-L2802

**Comment by zzrough** [comment]
Thanks for the very quick response and detailed explanation.

Le 11 novembre 2020 20:17:38 GMT+01:00, Ben Manes <notifications@github.com> a écrit :
>Thanks for the test case. It looks like this was optimized to use the
>backing CHM's faster `toArray`, but missed that it created an
>inconsistency by not checking for expiration and only iterating if
>`weakKeys` is enabled. The `values` and `entrySet` iterate, but since
>`keySet` might be optimizable it has this overload which contains that
>oversight. Guava always iterates, so we should be consistent.
>
>https://github.com/ben-manes/caffeine/blob/8c91f6544f8f7cbfa5275aa05cce0713ea7f2863/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L2792-L2802
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/ben-manes/caffeine/issues/472#issuecomment-725610749

--S

**Comment by zzrough** [comment]
wow both a quick response and a quick fix, amazing thanks a lot!

**Comment by ben-manes** [comment]
haha, yes. If this is frustrating then I can cut a release over the weekend. Just let me know.

**Comment by zzrough** [comment]
don't bother with an extra release if it's much work. I've worked around it simply using a collector like in the testcase.

Just release when you feel it's needed. I'll subscribe to the project release notifications.

Thanks again, have a great day.

**Comment by ben-manes** [comment]
Released 2.8.7
